### PR TITLE
lsp: document autocommand for enabling default signs and highlights

### LINKED ISSUE #12579: Changing Colorscheme causes highlight settings in diagnostics  to be cleared
Maybe we should think a little more about the call for this.
Changing color scheme causes the highlight setting in diagnostics to be cleared.

_Originally posted by @h-michael in https://github.com/neovim/neovim/pull/12333_

### PR DESCRIPTION
**Pull request by mjlbach**
@erw7 I agree with you that documenting this is the best option. I'm not sure if this is explicit enough though.

### CODE REVIEW

File: runtime/doc/lsp.txt [hunk 1]
```diff
@@ -425,6 +425,13 @@ therefore the linked highlights), use the |:highlight| command: >
     highlight LspDiagnosticsDefaultError guifg="BrightRed"
 <
 
+Note: Most custom colorschemes call `highlight clear` by convention on
+initialization. If the colorscheme does not define highlight groups for LSP
+diagnostics, you can set can set the highlights after the colorscheme has
+initialized with an autocommand: >
```
Comment (erw7) [inline runtime/doc/lsp.txt:431]: I think it would be better to specify that `autocommand` needs to be called before setting `colorshcme` as follows.

If the colorscheme does not define highlight groups for LSP diagnostics, you can set the highlights after the colorscheme has initialized with an autocommand (You will need to call autocommand before setting up colorschme):

Comment (mjlbach) [reply runtime/doc/lsp.txt]: Better?

File: runtime/doc/lsp.txt [hunk 2]
```diff
@@ -425,6 +425,14 @@ therefore the linked highlights), use the |:highlight| command: >
     highlight LspDiagnosticsDefaultError guifg="BrightRed"
 <
 
+Note: Most custom colorschemes call `highlight clear` by convention on
+initialization. If the colorscheme does not define highlight groups for
+LSP diagnostics, you can set the highlights after the colorscheme has
+initialized with an autocommand. You will need to set this autocommand
+before enabling the colorscheme: >
```
Comment (mhinz) [inline runtime/doc/lsp.txt:432]: If a blank line follows anyway, the `>` could be moved to the start of the next line. Just saying, no need to change it again. :)

The wording is clearer now.

**Review by mhinz — APPROVED** [review]

File: runtime/doc/lsp.txt [hunk 3]
```diff
@@ -425,6 +425,14 @@ therefore the linked highlights), use the |:highlight| command: >
...
+
+   autocmd ColorScheme * call v:lua.vim.lsp.diagnostic._define_default_signs_and_highlights()
```
Comment (lithammer) [inline runtime/doc/lsp.txt:434]: Does this function need to be promoted to public? Seems a bit weird to recommend using an internal function.

**Comment by jalvesaq** [comment]
I think that the lack of highlighting can mislead some people to think that Neovim's LSP is not ready for use yet and look for alternatives. As a user, my first step when software does not work well is to look for alternatives. I only read the documentation if I know that there is no alternative. I think most people do the same. In my opinion, since LSP is Neovim's built-in feature, it should have built-in highlighting. For example, if I comment out the line defining a highlighting group in my colorscheme (example, `String` or `Constant`), the group is highlighted with the same color that it would be highlighted if there was no colorscheme at all. But it's still highlighted.

In summary, if there are no colors, users will think that there is something wrong with the built-in LSP; if the colors do not match the current colorscheme, users will think that something is missing in the colorscheme.

I'm sorry if I missed something obvious, but I could only get diagnostic highlighted when I added the highlighting groups to my colorscheme.

**Comment by mjlbach** [comment]
@jalvesaq: Does adding `autocmd ColorScheme * call v:lua.vim.lsp.diagnostic._define_default_signs_and_highlights()` not fix it for you? We can't control what a random colorscheme plugin does

**Comment by jalvesaq** [comment]
@mjlbach, certainly adding the autocmd to `init.vim` will fix it (I didn't test). My point is that users will not and should not have to read the documentation to achieve simple goals such as getting things highlighted. So, in my opinion, at least the built-in color of `LspDiagnosticsDefaultError` should be the same as that of `Error` and that of `LspDiagnosticsDefaultWarning` should be the same as that of `WarningMsg`.

But perhaps I should open an issue requesting this feature because even if it was implemented it would not change the fact that user-defined highlightings will be cleared by most color schemes because they begin with the `hi clear` command, and your pull requests address this other problem.

**Comment by jalvesaq** [comment]
I opened an issue with my feature request: https://github.com/neovim/neovim/issues/14539

**Comment by gpanders** [comment]
Updated this for `vim.diagnostic`. The new method is

```vim
autocmd ColorScheme * source $VIMRUNTIME/plugin/diagnostic.vim
```

(I'm using `:source $VIMRUNTIME` rather than simply `:runtime` to avoid problems in the scenario that a user creates their own `~/.config/nvim/plugin/diagnostic.vim` file.)

**Comment by mjlbach** [comment]
@gpanders I can't approve my own PR but thanks for updating this haha (your commit is approved from my end).

>I wonder if we should add this as a default autocommand (#15401)? It's always safe to source this file since it uses :highlight default. The downside is that the file is then always sourced twice on startup (once when plugins are loaded and again on :colorscheme).

Mixed, I don't like the idea of sourcing a file twice.

**Comment by mjlbach** [comment]
Oh, actually we should go with #15894
